File: Tutorials/Codes/Robotics_Kinematics_and_Dynamics_Course/Chapter9/Lesson2/Chapter9_Lesson2.java
public class Planar2RVirtualWork {

    public static double[][] jacobian2R(double q1, double q2, double l1, double l2) {
        double s1 = Math.sin(q1);
        double c1 = Math.cos(q1);
        double s12 = Math.sin(q1 + q2);
        double c12 = Math.cos(q1 + q2);

        double[][] J = new double[2][2];
        J[0][0] = -l1 * s1 - l2 * s12;
        J[0][1] = -l2 * s12;
        J[1][0] =  l1 * c1 + l2 * c12;
        J[1][1] =  l2 * c12;
        return J;
    }

    public static double[] jointTorquesFromForce(double q1, double q2,
                                                 double l1, double l2,
                                                 double fx, double fy) {
        double[][] J = jacobian2R(q1, q2, l1, l2);
        double[] tau = new double[2];
        tau[0] = J[0][0]*fx + J[1][0]*fy;
        tau[1] = J[0][1]*fx + J[1][1]*fy;
        return tau;
    }

    public static void main(String[] args) {
        double q1 = Math.PI/4.0;
        double q2 = Math.PI/6.0;
        double l1 = 1.0, l2 = 0.8;
        double fx = 10.0, fy = 0.0;

        double[] tau = jointTorquesFromForce(q1, q2, l1, l2, fx, fy);
        System.out.println("tau1 = " + tau[0] + ", tau2 = " + tau[1]);
    }
}
      

File: Tutorials/Codes/Robotics_Kinematics_and_Dynamics_Course/Chapter9/Lesson2/Chapter9_Lesson2.py
import numpy as np

def jacobian_2r(q, l1, l2):
    q1, q2 = q
    s1 = np.sin(q1)
    c1 = np.cos(q1)
    s12 = np.sin(q1 + q2)
    c12 = np.cos(q1 + q2)

    J = np.array([
        [-l1*s1 - l2*s12, -l2*s12],
        [ l1*c1 + l2*c12,  l2*c12]
    ])
    return J

def joint_torques_from_force(q, l1, l2, f):
    """
    q  : array-like of shape (2,)  [q1, q2]
    l1,l2 : link lengths
    f  : array-like of shape (2,)  [fx, fy]
    returns tau : np.array shape (2,)
    """
    J = jacobian_2r(q, l1, l2)
    f_vec = np.asarray(f).reshape(2,)
    tau = J.T @ f_vec
    return tau

if __name__ == "__main__":
    q = np.deg2rad([45.0, 30.0])
    l1, l2 = 1.0, 0.8
    f = np.array([10.0, 0.0])  # force in x-direction
    tau = joint_torques_from_force(q, l1, l2, f)
    print("Joint torques:", tau)
      
